Remove cancellation tokens from FAR context objects.

CyrusNajmabadi opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/51045
Fixes https://github.com/dotnet/roslyn/issues/52810

Current design of FAR has us pass around context objects that contain a cancellation token in it representing the cancellability of the FAR operation.  THis approach has led to multiple bugs (and has necessitated other fixes like https://github.com/dotnet/roslyn/pull/50948) where different parts of hte code disagree on the source of truth of cancellability.  This is particularly problematic in OOP which has to coordinate cancellation between two separate processes and the messages going to and from them in a streaming context.  

This PR undoes how we do cancellation in FAR entirely.  The value is removed from our context object, and we just pass along a CT as a parameter like we do everywhere else.  This was *extremely* viral, but should put us in a clean state from now on.

CyrusNajmabadi commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindBaseSymbolsCommandHandler.cs:
```diff
@@ -66,31 +66,32 @@ protected override bool TryExecuteCommand(int caretPosition, Document document,
 
-                var context = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
+                var (context, combinedCancellationToken) = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
+                cancellationToken = combinedCancellationToken;
```
alternatively, we could consider passing the cancellation token by-ref to StartSearch.  However, the amount of code that calls StartSearch is minimal, so this pattern works.

CyrusNajmabadi commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindBaseSymbolsCommandHandler.cs:
```diff
@@ -66,31 +66,32 @@ protected override bool TryExecuteCommand(int caretPosition, Document document,
 #pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task
-                        var relevantSymbol = await FindUsagesHelpers.GetRelevantSymbolAndProjectAtPositionAsync(document, caretPosition, context.CancellationToken);
+                        var relevantSymbol = await FindUsagesHelpers.GetRelevantSymbolAndProjectAtPositionAsync(document, caretPosition, cancellationToken);
 #pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
```
i have no clue why none of the cocoa code avoids calling ConfigureAwait.  @KirillOsenkov do you know why?

KirillOsenkov replied:
I have no idea. It was originally written like this by @avodovnik and then @tmat added the pragmas in a bulk commit to clean up warnings: https://github.com/dotnet/roslyn/commit/589da9de5ec3e2bd2a58d727c92ea76978ffaa0e

sharwell replied:
ConfigureAwait is generally discouraged in JTF-aware code.

CyrusNajmabadi replied:
so i don't think this code is JTF aware.  My preference would just be to remove the pragmas and `ConfigureAwait(false) ` all this code as it feels safe to do so.  Waiting on @KirillOsenkov though to determine if that's the case.

CyrusNajmabadi commented on src/Features/Core/Portable/FindUsages/IRemoteFindUsagesService.cs:
```diff
@@ -92,54 +91,40 @@ public FindUsagesServerCallback(Solution solution, IFindUsagesContext context)
 
-                await _context.OnDefinitionFoundAsync(rehydrated).ConfigureAwait(false);
-            }
-            catch (OperationCanceledException oce) when (oce.CancellationToken == _context.CancellationToken)
```
an example of wonkyness that we had code to try to be resilient around.  this should be unnecessary now as tehre are no shenanigans around CTs.

CyrusNajmabadi commented:
Bleagh.  Just realized this will impact TypeScript.  I need to create an EA layer for them.  Working on that now.

CyrusNajmabadi commented on src/EditorFeatures/Core/FindUsages/IFindUsagesService.cs:
```diff
@@ -2,59 +2,25 @@
 
 namespace Microsoft.CodeAnalysis.Editor.FindUsages
 {
-    [Obsolete("Legacy API for TypeScript.  Once TypeScript moves to IVSTypeScriptFindUsagesService", error: false)]
```
typescript has moved to IVSTypeScriptFindUsagesService, so we do not need the bridge interface.

sharwell commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindDerivedSymbolsCommandHandler.cs:
```diff
@@ -91,37 +91,38 @@ private static async Task<IEnumerable<ISymbol>> GatherSymbolsAsync(ISymbol symbo
                 using var token = _asyncListener.BeginAsyncOperation(nameof(FindDerivedSymbolsAsync));
 
-                var context = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
+                var (context, combinedCancellationToken) = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
+                cancellationToken = combinedCancellationToken;
```
💭 I believe this might now be supported?

```suggestion
                (var context, cancellationToken) = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
```

CyrusNajmabadi replied:
ah, i'll try that.  i hate it.  but i hate it less than the current code.

CyrusNajmabadi replied:
much nicer.  thanks!

sharwell commented on src/VisualStudio/Core/Def/Implementation/FindReferences/StreamingFindUsagesPresenter.cs:
```diff
@@ -176,6 +176,10 @@ public FindUsagesContext StartSearch(string title, bool supportsReferences, Canc
 
+            // Wrap the passed in CT with our own CTS that we can control cancellation over.  This way either our
+            // caller can cancel our work or we can cancel the work.
+            var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```
❓ Where does this get disposed (cancellation token sources created by `CreateLinkedTokenSource` must be disposed or they will leak memory)

CyrusNajmabadi replied:
ah.  the presenter is disposable.  i will add.  thanks!

CyrusNajmabadi replied:
so we do properly already dispose thsi.  i will add comment here about the ownership going on here.

sharwell commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine_MapCreation.cs:
```diff
@@ -145,48 +147,49 @@ private async Task<ProjectMap> CreateProjectMapAsync(ConcurrentSet<SymbolGroup>
                 {
                     Contract.ThrowIfNull(symbol);
-                    _cancellationToken.ThrowIfCancellationRequested();
+                    cancellationToken.ThrowIfCancellationRequested();
```
⚠️ In the event of cancellation,  the caller still needs to wait for tasks in `symbolTasks` to complete

CyrusNajmabadi replied:
not sure i understand this comment.  can you clarify?

sharwell replied:
At the point this exception is thrown, `symbolTasks` might contain one or more tasks which has not been awaited (not yet complete). The caller needs to wait for these tasks to complete as part of its cancellation handling.

CyrusNajmabadi replied:
I see. Will address

CyrusNajmabadi replied:
done.  this was a very good catch.  i'm wondering if this sort of mistake could be made easier to detect...

jasonmalinowski reviewed (commented):
(posting comments so far -- wasn't able to make it through all the way)

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine_DocumentProcessing.cs:
```diff
@@ -15,26 +16,27 @@ internal partial class FindReferencesSearchEngine
         private async Task ProcessDocumentQueueAsync(
             Document document,
-            HashSet<(SymbolGroup group, ISymbol symbol, IReferenceFinder finder)> documentQueue)
+            HashSet<(SymbolGroup group, ISymbol symbol, IReferenceFinder finder)> documentQueue,
```
they are all on separate lines... :)

CyrusNajmabadi commented on src/EditorFeatures/Core/FindUsages/FindUsagesContext.cs:
```diff
@@ -12,26 +10,21 @@ namespace Microsoft.CodeAnalysis.FindUsages
 
-        ValueTask IFindUsagesContext.ReportProgressAsync(int current, int maximum)
-            => ReportProgressAsync(current, maximum);
+        protected virtual ValueTask ReportProgressAsync(int current, int maximum, CancellationToken cancellationToken) => default;
```
happy to explain more tomorrow. however, here's twhat's going on (which seems pretty reasonable):

1. the ProgressTracker is what the context externally exposes.  It's how features like FAR itself say "here's the progress i'm making"
2. the progress tracker works by internally keeping track of state, but then calling out to a callback whenever its state changes.  The callback in this case is hte 'ReportProgressAsync' method.
3. The actual FAR window (which provides a subclass of this context) overrides ReportProgresAsync to actually update the UI.

> The context is handed to a feature to actually do the work, which can only call ReportProgressAsync

No.  The context is handed to a feature, which uses ProgressTracker.  

> But we have a protected ProgressTracker

No.  ProgressTracker is not protected.  It's public and is what consumers are intended to use to report progress.

> which will also call ReportProgress and I presume stomp on each other,

No.  The ProgressTracker calls ReportProgressAsync as a way of telling whoever is the context: update your UI to reflect the state of the progress however hte feature has asked us to.

LMK if you need anymore info o nthis.

jasonmalinowski replied:
Oh I'm sorry I totally misread the diff: I saw ReportProgressAsync still being left on the interface, but it was ReportMessageAsync. Got it now. Just too much happening at the same time and the diff is getting hard to read.

jasonmalinowski commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindBaseSymbolsCommandHandler.cs:
```diff
@@ -66,31 +66,31 @@ protected override bool TryExecuteCommand(int caretPosition, Document document,
                 using var token = _asyncListener.BeginAsyncOperation(nameof(StreamingFindBaseSymbolsAsync));
 
-                var context = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
+                (var context, cancellationToken) = presenter.StartSearch(EditorFeaturesResources.Navigating, supportsReferences: true, cancellationToken);
```
So this pattern at first to me looked like an outright bug -- the fact we're now combining with a new cancellation token meant that this method could throw cancellation which was _not_ triggered by the original requestor, which is always a no-no. But it's even funkier: the caller is passing in CancellationToken.None!

Note this is not a bug being introduced by this PR -- this was previously broken, the mistake is now much more obvious. The fix would be to change this pattern a bit: the call to presenter.StartSearch should be moved back into TryExecuteCommand, and you can pass the cancellation token from the CommandExecutionContext. From there, Task.Run(StreamingFindBaseSymbolsAsync), and there you can pass the cancellation token you receive from StartSearch -- essentially think of it as:

1. During the initial starting of the search, any delay is only cancellable by the threaded wait dialog, since we don't have another UI affordance for the user to do anything else.
2. Once the presenter is up, there's a cancel button there, and that's what the user uses.

jasonmalinowski replied:
(note this comment applies to the other command handlers in this folder, since the same pattern/problem applies to all)

CyrusNajmabadi replied:
yup.  this has been fixed.

jasonmalinowski commented on src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs:
```diff
@@ -34,14 +34,14 @@ internal interface IStreamingFindUsagesPresenter
         /// <param name="cancellationToken">External cancellation token controlling whether finding shoudl be canceled
         /// or not.  This will be combined with a cancellation token owned by the <see cref="FindUsagesContext"/>.
-        /// Callers should consider <see cref="FindUsagesContext.CancellationToken"/> to be the source of truth for
+        /// Callers should consider the cancellation token returned to be the source of truth for
         /// cancellation from that point on.</param>
```
See comments elsewhere: this is a pretty dangerous pattern since it means callers may throw cancellation when the original caller didn't expect it. I wonder if that cancellation token confusion is partially the cause of mistakes here, and we aren't rectifying it. My gut is saying that starting a search isn't a cancellable operation  -- we're just opening up a dialog. The caller either never got to that point, or has chosen to show a dialog at which point the original cancellation is now useless -- there's no more waiting, and we're now only relying on some external cancellation UI.....right?

CyrusNajmabadi replied:
resolved.

jasonmalinowski reviewed (commented):
Pausing review since I'm still seeing a lot of potential cancellation token mishandling. I'm not saying this PR _doesn't_ fix the original bugs, but I'm strongly wondering if fact that StartSearch takes a cancellation token and the lack of care around callers dealing with the fact there are two might be a root problem here, and I'm not sure this PR makes that part better.

CyrusNajmabadi commented:
Addressed feedback from Jason.  The idea is this:

1. we are nto combining cancellation tokens anymore.  That's just not what we want.
2. Instead, we can have two cancellation tokens, and which we use depends on hte scenario.  If the feature is effectively blocking, and is just using the window as a place to put its results into, then we should use the cancelaltion token that feature has, as that's what will be in the TWD the user can use to interrupt the work.
3. However, if teh feature is non-blocking (like normal FAR is), then we explicitly do not use the cancellation token of the command handler (as that is only relevant while the command is initially dispatched).  Once it's converted into fire-and-forget work happening in the background, then at that point the cancellation token of the window takes over.  This cancellation token only gets fired then if the window is closed, or a new search happens which takes it over.

CyrusNajmabadi commented:
@KirillOsenkov have tagged you to look at teh cocoa parts (which i think are good).  Also, tehre are a lot of supressions in teh cocoa part about using `ConfigureAwait(true/false)`.  Are they necessary?  I imagine it should be fine to remove those and add in COnfigureAwait(false) here as i don't imagine your code has any UI dependencies.  Note that if it does have UI dependencies, they can go in your impl of the IStreamingUsagesPresenter window as all the callbacks are async, so you can always jump to teh UI thread then.  Does that work for you?

KirillOsenkov commented:
I suppose so. I have no idea why we would want to avoid ConfigureAwait.

CyrusNajmabadi commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindBaseSymbolsCommandHandler.cs:
```diff
@@ -49,8 +48,7 @@ protected override bool TryExecuteCommand(int caretPosition, Document document,
             {
-                // Fire and forget.  So no need for cancellation.
-                _ = StreamingFindBaseSymbolsAsync(document, caretPosition, streamingPresenter, CancellationToken.None);
+                _ = StreamingFindBaseSymbolsAsync(document, caretPosition, streamingPresenter);
```
all the cocoa code operates ina  fire-and-forget manner, where the ui thread is relinguished back to the user, and the results are streamed into the FAR window.  AS such, we do *not* pass a cancellation token along here.  And in teh fire-and-forget method, we grab the CT from teh presenter and use that to control teh search.

jasonmalinowski commented on src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs:
```diff
@@ -32,16 +32,18 @@ internal interface IStreamingFindUsagesPresenter
-        FindUsagesContext StartSearchWithCustomColumns(string title, bool supportsReferences, bool includeContainingTypeAndMemberColumns, bool includeKindColumn, CancellationToken cancellationToken);
+        (FindUsagesContext context, CancellationToken cancellationToken) StartSearchWithCustomColumns(string title, bool supportsReferences, bool includeContainingTypeAndMemberColumns, bool includeKindColumn);
 
         /// <summary>
```
Did you want to delete ClearAll since we're already here and breaking the interface anyways?

CyrusNajmabadi replied:
@KirillOsenkov can you check if you guys use `ClearAll` at all in VS4mac for us?  thanks!

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs:
```diff
@@ -228,11 +222,11 @@ private void CancelSearch()
-                    _cancellationTokenSource = null;
+                    CancellationTokenSource.Cancel();
+                    CancellationTokenSource.Dispose();
+                    CancellationTokenSource = null;
```
I'm confused by the null-out here: the underlying data the CancellationTokenSource holds onto is gotten rid of in the Dispose() case, right? The docs say things like "you must release all references" but I can't tell if that's just automatic stuff the docs generate for all IDisposable things. Specifically this warning:

> Always call Dispose before you release your last reference to the CancellationTokenSource. Otherwise, the resources it is using will not be freed until the garbage collector calls the CancellationTokenSource object's Finalize method.

seems off because the implementation verbatim says:

> // There is nothing to do if disposing=false because the CancellationTokenSource holds no unmanaged resources.

@stephentoub is this null-out actually required, or is this just something the docs generate automatically for those who think IDisposable.Dispose() also makes the direct object go away?

CyrusNajmabadi replied:
> Otherwise, the resources it is using will not be freed until the garbage collector calls the CancellationTokenSource object's Finalize method.

Honestly, if this is the case, I'd be 100% ok not disposing then at all. It just adds complexity to lots of scenarios that we could avoid if we know the GC will just handle this for us

CyrusNajmabadi replied:
>  it is using will not be freed until the garbage collector calls the CancellationTokenSource object's Finalize method.

Hrmm... i took a look, and CTS doesn't even have a finalizer afaict... So i'm not sure if these comments are valid.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/FindReferences/StreamingFindUsagesPresenter.cs:
```diff
@@ -137,38 +137,28 @@ public void ClearAll()
             // this set.
             _currentContexts.Add(context);
-            return context;
+            return (context, context.CancellationTokenSource!.Token);
```
See comment earlier, I'm not sure why CancellationTokenSource is nullable.

CyrusNajmabadi replied:
In this case, I think it's too make sure we don't accidentally double dispose. If it's safe to do that, then we can remove the null out

jasonmalinowski reviewed (approved):
:shipit:

We still need to clean up the VS for Mac stuff obviously, and I'm confused by the docs on CancellationTokenSource but neither is blocking. Obviously please test this carefully with TypeScript/F# to prevent IVT breaks.

(this more generally also reminds me that we now have so many layers of find references and progress abstractions , but alas...)

CyrusNajmabadi commented on src/EditorFeatures/Core.Cocoa/NavigationCommandHandlers/FindBaseSymbolsCommandHandler.cs:
```diff
@@ -59,38 +57,37 @@ protected override bool TryExecuteCommand(int caretPosition, Document document,
                     FunctionId.CommandHandler_FindAllReference,
                     KeyValueLogMessage.Create(LogType.UserAction, m => m["type"] = "streaming"),
-                    context.CancellationToken))
+                    cancellationToken))
```
taht's fine.  because these are all fire-and-forget async methods, they all have an appropriate:

```c#
try
{
    await the async work
}
catch (OperationCanceled)
{
}
catch (Exception ex) when (FatalError.ReportAndCatch(ex))
{
}
```

this makes sure no exceptions leak out, and that we properly NFW report the issues that may arise.

CyrusNajmabadi commented on src/EditorFeatures/Core/FindReferences/FindReferencesCommandHandler.cs:
```diff
@@ -123,25 +120,24 @@ private bool TryExecuteCommand(int caretPosition, Document document, IFindUsages
                     try
                     {
-                        await findUsagesService.FindReferencesAsync(document, caretPosition, context).ConfigureAwait(false);
+                        await findUsagesService.FindReferencesAsync(document, caretPosition, context, cancellationToken).ConfigureAwait(false);
```
see above comment. fire-and-forget methods have a try/catch in them that explicitly swallows these.
